staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -67,36 +69,81 @@ type cacheEntry struct {
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+
+	// Key already exists
+	oldElement, ok := c.entries[key]
+	if ok {
+		c.evictionList.MoveToFront(oldElement)
+		oldElement.Value.(*cacheEntry).value = value
+		oldElement.Value.(*cacheEntry).expireTime = c.clock.Now().Add(ttl)
+		return
+	}
+
+	// Make space if necessary
+	if c.evictionList.Len() >= c.maxSize {
+		toEvict := c.evictionList.Back()
+		c.evictionList.Remove(toEvict)
+		delete(c.entries, toEvict.Value.(*cacheEntry).key)
```

Comment:
a maxSize < 0 would run this block when c.evictionList was empty, toEvict would be nil, and `toEvict.Value.(*cacheEntry).key` would hit a NPE.

keeping the panic in NewLRUExpireCacheWithClock if maxSize is < 0 would surface that coding problem sooner

Reply:
Done, though I panic on maxSize <= 0, since I think 0 is also a problematic value here.